lib/instrumentation/http-shared.js

```diff
@@ -152,23 +153,25 @@ exports.traceOutgoingRequest = function (agent, moduleName, method) {
       // If so, then fallback to the current run context's span or transaction,
       // if any.
       var parent = span || parentRunContext.currSpan() || parentRunContext.currTransaction()
-      if (parent && parent._context) {
-        const headerValue = parent._context.toTraceParentString()
-        const traceStateValue = parent._context.toTraceStateString()
-
-        const newHeaders = Object.assign({}, options.headers)
-        newHeaders.traceparent = headerValue
-        if (traceStateValue) {
-          // Do not set the header to the empty string. Though RFC 7230 allows
-          // empty header values, some servers apparently do not.
-          newHeaders.tracestate = traceStateValue
+
+      const setter = function (carrier, name, value) {
+        if (!value) {
+          return
         }
-        if (agent._conf.useElasticTraceparentHeader) {
-          newHeaders['elastic-apm-traceparent'] = headerValue
+        carrier[name] = value
+
+        if (agent._conf.useElasticTraceparentHeader && name === 'traceparent') {
+          carrier['elastic-apm-traceparent'] = value
         }
```

Comment:
The setter should be called by TraceContext for each header it wants to set. So instead of hard-coding knowledge about the header names in each setter, can this if block be part of TraceContext?

Reply:
Seems reasonable -- (I think I avoided doing this because I tend to think of this specific header as a legacy thing that we'll eventually not want to carry forward _and_ it can sometimes be tricky for us to get the agent conf injected into the place we need to check a config) done in 7cbbec6c64f2db5393c46f83c20dcd7373d689ef